ding/policy/prompt_pg.py

```diff
@@ -0,0 +1,255 @@
+                act.append(actions)
+                for jj in range(actions.shape[0]):
+                    mask[jj][actions[jj]] = -1e30
+            # `act` is shaped (shot_num, B)
+            real_act = []
+            for b in range(act[0].shape[0]):
+                tmp_act = torch.zeros_like(act[0])
+                for shot in act:
+                    tmp_act += 2 ** shot[b].item()
+                    real_act.append(tmp_act)
+            real_act = torch.tensor(real_act)
+            # `real_act` is shaped (B)
+        output['action'] = real_act
+        if self._cuda:
+            output = to_device(output, 'cpu')
+        output = default_decollate(output)
+        return {i: d for i, d in zip(data_id, output)}
+
+    def _process_transition(self, obs: Any, model_output: dict, timestep: namedtuple) -> dict:
+        r"""
+        Overview:
+            Generate dict type transition data from inputs.
+        Arguments:
+            - obs (:obj:`Any`): Env observation
+            - model_output (:obj:`dict`): Output of collect model, including at least ['action']
+            - timestep (:obj:`namedtuple`): Output after env step, including at least ['obs', 'reward', 'done'] \
+                (here 'obs' indicates obs after env step).
+        Returns:
+            - transition (:obj:`dict`): Dict type transition data.
+        """
+        return {
+            'obs': obs,
+            'action': model_output['action'],
+            'reward': timestep.reward,
+            'done': timestep.done,
+        }
+
+    def _get_train_sample(self, data: list) -> Union[None, List[Any]]:
+        r"""
+        Overview:
+            Get the trajectory and the n step return data, then sample from the n_step return data
+        Arguments:
+            - data (:obj:`list`): The trajectory's buffer list
+        Returns:
+            - samples (:obj:`dict`): The training samples generated
+        """
+        if self._cfg.learn.ignore_done:
+            raise NotImplementedError
+
+        R = 0.
+        for i in reversed(range(len(data))):
+            R = self._gamma * R + data[i]['reward']
+            data[i]['return'] = R
+        return get_train_sample(data, self._unroll_len)
+
+    def _init_eval(self) -> None:
+        pass
+
+    def _forward_eval(self, data: dict) -> dict:
+        data_id = list(data.keys())
+        data = default_collate(list(data.values()))
+        self._model.eval()
+        with torch.no_grad():
+            # Prepare train_sample (the question to be answered) and the candidate_samples (the prompts to be selected)
+            for ii in range(len(data['candidate_samples'])):
+                data['candidate_samples'][ii] = data['candidate_samples'][ii][0]
+            output = self._model.forward(data['train_sample'], data['candidate_samples'])
+            # Generate actions.
+            act = []
+            mask = torch.zeros_like(output['logit'])
+            for ii in range(self._cfg.shot_number):
```

Comment:
maybe we can abstarct the operation of the shot_number into a new head in `ding/model/common`